src/bin/render-readmes.rs

```diff
@@ -2,87 +2,146 @@
+        .filter(readme_rendering::rendered_at.lt(older_than))
+        .select(versions::id)
+        .into_boxed();
+
+    if let Some(crate_name) = args.flag_crate {
+        println!("Rendering readmes for {}", crate_name);
+        query = query.filter(crates::name.eq(crate_name));
+    }
+
+    let version_ids = query.load::<(i32)>(&conn).expect(
+        "error loading version ids",
+    );
+
+    let total_versions = version_ids.len();
+    println!("Rendering {} versions", total_versions);
+
+    let page_size = args.flag_page_size.unwrap_or(DEFAULT_PAGE_SIZE);
+
+    let total_pages = total_versions / page_size;
+    let total_pages = if total_versions % page_size == 0 {
+        total_pages
     } else {
-        versions_count / page_size + 1
+        total_pages + 1
     };
-    for current_page in 0..pages {
-        let versions: Vec<EncodableVersion> = versions::table
+
+    let mut page_num = 0;
+
+    for version_ids_chunk in &version_ids.into_iter().chunks(page_size) {
+        page_num += 1;
+        println!(
+            "= Page {} of {} ==================================",
+            page_num,
+            total_pages
+        );
+
+        let ids: Vec<_> = version_ids_chunk.collect();
+
+        let versions = versions::table
             .inner_join(crates::table)
+            .filter(versions::id.eq(any(ids)))
             .select((versions::all_columns, crates::name))
-            .limit(page_size)
-            .offset(current_page * page_size)
             .load::<(Version, String)>(&conn)
-            .expect("error loading versions")
-            .into_iter()
-            .map(|(version, crate_name)| version.encodable(&crate_name))
-            .collect();
+            .expect("error loading versions");
+
         let mut tasks = Vec::with_capacity(page_size as usize);
-        for version in versions {
+        for (version, krate_name) in versions {
             let config = config.clone();
+            version.record_readme_rendering(&conn).expect(&format!(
+                "[{}-{}] Couldn't record rendering time",
+                krate_name,
+                version.num
+            ));
             let handle = thread::spawn(move || {
-                println!("[{}-{}] Rendering README...", version.krate, version.num);
-                let readme = get_readme(&config, &version);
+                println!("[{}-{}] Rendering README...", krate_name, version.num);
+                let readme = get_readme(&config, &version, &krate_name);
                 if readme.is_none() {
                     return;
                 }
                 let readme = readme.unwrap();
-                let readme_path = format!(
-                    "readmes/{}/{}-{}.html",
-                    version.krate,
-                    version.krate,
-                    version.num
-                );
+                let readme_path =
+                    format!("readmes/{}/{}-{}.html", krate_name, krate_name, version.num);
```

Comment:
Should we change the formatting string to `"readmes/{0}/{0}-{}.html"` to avoid the `krate_name` duplication ?

Reply:
sure!